src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,143 @@
+ * The attribute value encoder can be initialized with a AttributeEncodeState for saving and recovering its state between encode
+ * sessions (chunkings).
+ *
+ * When Encode returns recoverable errors (e.g. CHIP_ERROR_NO_MEMORY) the state can be used to initialize the AttributeValueEncoder
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
+{
+public:
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
+    };
+
+    /**
+     * AttributeReportBuilder offleads some code in AttributeValueEncoder::Encode to keep its content small.
+     */
+    AttributeValueEncoder(AttributeReportIBs::Builder & aAttributeReportIBsBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, DataVersion aDataVersion,
+                          const AttributeEncodeState & aState = AttributeEncodeState()) :
+        mAttributeReportIBsBuilder(aAttributeReportIBsBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mDataVersion(aDataVersion), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+        AttributeReportBuilder builder;
+
+        ReturnErrorOnFailure(builder.PrepareAttribute(
+            mAttributeReportIBsBuilder,
+            ConcreteDataAttributePath(mPath.mEndpointId, mPath.mClusterId, mPath.mAttributeId,
+                                      // We only use the two operations here to indicate if we need to encode a null list index
+                                      mCurrentEncodingListIndex == kInvalidListIndex
+                                          ? ConcreteDataAttributePath::ListOperation::NotList
+                                          : ConcreteDataAttributePath::ListOperation::AppendItem,
+                                      mCurrentEncodingListIndex),
+            mDataVersion));
+        ReturnErrorOnFailure(builder.Encode(std::forward<Ts>(aArgs)...));
+
+        return builder.FinishAttribute();
+    }
+
+    template <typename... Ts>
+    CHIP_ERROR EncodeListItem(Ts... aArgs)
```

Comment:
What assumptions is this making about mCurrentEncodingListIndex and mEncodeState.mCurrentEncodingListIndex?  Incrementing them if they're kInvalidListIndex seems odd.  Is the assumption that they are not?  Why not?  What ensures that?

Or put another way: How does the consumer know they're supposed to call this function?